src/compose/app.ts

```diff
@@ -307,42 +307,43 @@ export class App {
 			}
 		}
 
-		const alreadyStarted = (serviceId: number) => {
-			const equalExceptForRunning = currentByServiceId[
-				serviceId
-			].isEqualExceptForRunningState(
-				targetByServiceId[serviceId],
-				containerIds,
-			);
-			if (!equalExceptForRunning) {
-				// We need to recreate the container, as the configuration has changed
-				return false;
-			}
-
-			if (targetByServiceId[serviceId].config.running) {
-				// If the container is already running, and we don't need to change it
-				// due to the config, we know it's already been started
+		const alreadyStarted = (service: Service) => {
+			// If service is currently running then we started it previously
+			if (service.config.running) {
 				return true;
 			}
+			// Lastly, check if we previously remember starting it
+			return applicationManager.containerStarted[service.containerId!] != null;
+		};
 
-			// We recently ran a start step for this container, it just hasn't
-			// started running yet
+		const runningStatesMatch = (serviceId: number) => {
 			return (
-				applicationManager.containerStarted[
-					currentByServiceId[serviceId].containerId!
-				] != null
+				currentByServiceId[serviceId].config.running ===
+				targetByServiceId[serviceId].config.running
 			);
 		};
 
+		const doNotRestart = (service: Service) => {
+			return service.config.restart === '';
+		};
+
 		const needUpdate = maybeUpdate.filter(
+			// Needs an update if NOT...
 			(serviceId) =>
 				!(
-					currentByServiceId[serviceId].isEqual(
+					alreadyStarted(currentByServiceId[serviceId]) && // 1 - Been started before
+					// 2 - Running states match OR Service stopped AND Supervisor should not restart
+					(runningStatesMatch(serviceId) ||
+						(!currentByServiceId[serviceId].config.running &&
+							doNotRestart(currentByServiceId[serviceId]))) &&
```

Comment:
@20k-ultra I really don't think this is a check we should be doing; if the `alreadyStarted` check considers if it's been started in the past then this check is moot and just adds to the complexity. What am I missing?

Reply:
confirmed we don't to check restart policy.